File: products/sdlc-cc/platform/services/rag/app/services/opa_client_test.py
"""
Tests for OPA Client Service
"""

import asyncio
import json
from datetime import datetime
from unittest.mock import AsyncMock, patch

import pytest
from aioresponses import aioresponses
from pydantic import ValidationError

from .opa_client import (
    OPAClient,
    OPAConfig,
    PolicyEvaluationResponse,
    BatchEvaluation,
    OPATimeoutError,
    OPAApiError,
    create_opa_client,
)


class TestOPAConfig:
    """Test OPA configuration"""

    def test_default_config(self):
        """Test default configuration values"""
        config = OPAConfig()
        assert config.base_url == "http://localhost:8181"
        assert config.timeout == 5
        assert config.cache_enabled is True
        assert config.cache_ttl == 30
        assert config.retry_attempts == 3
        assert config.retry_delay == 0.1

    def test_custom_config(self):
        """Test custom configuration"""
        config = OPAConfig(
            base_url="http://custom:8181", timeout=10, cache_enabled=False, cache_ttl=60
        )
        assert config.base_url == "http://custom:8181"
        assert config.timeout == 10
        assert config.cache_enabled is False
        assert config.cache_ttl == 60

    def test_invalid_config(self):
        """Test invalid configuration validation"""
        with pytest.raises(ValidationError):
            OPAConfig(timeout=-1)


class TestOPAClient:
    """Test OPA client functionality"""

    @pytest.fixture
    def mock_redis(self):
        """Mock Redis client"""
        redis_mock = AsyncMock()
        redis_mock.get = AsyncMock(return_value=None)
        redis_mock.setex = AsyncMock(return_value=True)
        redis_mock.delete = AsyncMock(return_value=1)
        return redis_mock

    @pytest.fixture
    def opa_client(self, mock_redis):
        """Create OPA client instance"""
        config = OPAConfig(cache_enabled=True, timeout=1)
        return OPAClient(config=config, redis_client=mock_redis)

    @pytest.fixture
    def sample_opa_response(self):
        """Sample OPA response"""
        return {"result": {"allow": True, "decision_reason": ["Access granted"]}}

    @pytest.mark.asyncio
    async def test_evaluate_policy_success(self, opa_client, sample_opa_response):
        """Test successful policy evaluation"""
        with aioresponses() as m:
            m.post(
                "http://localhost:8181/v1/data/test.policy",
                payload=sample_opa_response,
                status=200,
            )

            async with opa_client:
                response = await opa_client.evaluate_policy(
                    "test.policy", {"user": "test", "action": "read"}
                )

            assert response.decision is True
            assert response.reason == "Access granted"
            assert response.result == sample_opa_response["result"]

    @pytest.mark.asyncio
    async def test_evaluate_policy_deny(self, opa_client):
        """Test policy evaluation with deny decision"""
        response_data = {
            "result": {"allow": False, "decision_reason": ["Access denied"]}
        }

        with aioresponses() as m:
            m.post(
                "http://localhost:8181/v1/data/test.policy",
                payload=response_data,
                status=200,
            )

            async with opa_client:
                response = await opa_client.evaluate_policy(
                    "test.policy", {"user": "test", "action": "delete"}
                )

            assert response.decision is False
            assert response.reason == "Access denied"

    @pytest.mark.asyncio
    async def test_evaluate_policy_api_error(self, opa_client):
        """Test policy evaluation with API error"""
        with aioresponses() as m:
            m.post(
                "http://localhost:8181/v1/data/test.policy",
                payload={"error": "policy not found"},
                status=404,
            )

            async with opa_client:
                with pytest.raises(OPAApiError) as exc_info:
                    await opa_client.evaluate_policy("test.policy", {"user": "test"})

            assert exc_info.value.status_code == 404

    @pytest.mark.asyncio
    async def test_evaluate_policy_timeout(self, opa_client):
        """Test policy evaluation timeout"""
        with aioresponses() as m:
            m.post(
                "http://localhost:8181/v1/data/test.policy",
                exception=asyncio.TimeoutError(),
            )

            async with opa_client:
                with pytest.raises(OPATimeoutError):
                    await opa_client.evaluate_policy("test.policy", {"user": "test"})

    @pytest.mark.asyncio
    async def test_evaluate_policy_retry(self, opa_client):
        """Test policy evaluation with retry"""
        with aioresponses() as m:
            # First attempt fails
            m.post(
                "http://localhost:8181/v1/data/test.policy",
                exception=Exception("Network error"),
            )
            # Second attempt succeeds
            m.post(
                "http://localhost:8181/v1/data/test.policy",
                payload={"result": {"allow": True}},
                status=200,
            )

            config = OPAConfig(retry_attempts=2, retry_delay=0.01)
            client = OPAClient(config=config, redis_client=opa_client.redis)

            async with client:
                response = await client.evaluate_policy("test.policy", {"user": "test"})

            assert response.decision is True

    @pytest.mark.asyncio
    async def test_evaluate_data_policy(self, opa_client, sample_opa_response):
        """Test data policy evaluation"""
        with aioresponses() as m:
            m.post(
                "http://localhost:8181/v1/data/sdlc.data.access",
                payload=sample_opa_response,
                status=200,
            )

            async with opa_client:
                response = await opa_client.evaluate_data_policy(
                    tenant_id="tenant-123",
                    user_id="user-123",
                    action="read",
                    resource="documents",
                    data={"id": "doc-123"},
                )

            assert response.decision is True

    @pytest.mark.asyncio
    async def test_evaluate_auth_policy(self, opa_client, sample_opa_response):
        """Test authentication policy evaluation"""
        with aioresponses() as m:
            m.post(
                "http://localhost:8181/v1/data/sdlc.auth.policy",
                payload=sample_opa_response,
                status=200,
            )

            async with opa_client:
                response = await opa_client.evaluate_auth_policy(
                    {"user_id": "user-123", "role": "admin"}
                )

            assert response.decision is True

    @pytest.mark.asyncio
    async def test_evaluate_dlp_policy(self, opa_client):
        """Test DLP policy evaluation"""
        response_data = {
            "result": {
                "allow": False,
                "decision_reason": ["Content blocked: contains high-risk PII"],
            }
        }

        with aioresponses() as m:
            m.post(
                "http://localhost:8181/v1/data/sdlc.dlp.policy",
                payload=response_data,
                status=200,
            )

            async with opa_client:
                response = await opa_client.evaluate_dlp_policy(
                    content="SSN: 123-45-6789",
                    user_context={"user_id": "user-123", "role": "user"},
                )

            assert response.decision is False
            assert "high-risk PII" in response.reason

    @pytest.mark.asyncio
    async def test_evaluate_rag_policy(self, opa_client, sample_opa_response):
        """Test RAG policy evaluation"""
        with aioresponses() as m:
            m.post(
                "http://localhost:8181/v1/data/sdlc.rag.policy",
                payload=sample_opa_response,
                status=200,
            )

            async with opa_client:
                response = await opa_client.evaluate_rag_policy(
                    query="What is the revenue?",
                    user_context={"user_id": "user-123", "role": "analyst"},
                    documents=[{"id": "doc-1", "content": "Revenue is $1M"}],
                )

            assert response.decision is True

    @pytest.mark.asyncio
    async def test_batch_evaluate_policies(self, opa_client):
        """Test batch policy evaluation"""
        with aioresponses() as m:
            for i in range(3):
                m.post(
                    f"http://localhost:8181/v1/data/test.policy.{i}",
                    payload={"result": {"allow": True}},
                    status=200,
                )

            async with opa_client:
                evaluations = [
                    BatchEvaluation("test.policy.0", {"action": "read"}),
                    BatchEvaluation("test.policy.1", {"action": "write"}),
                    BatchEvaluation("test.policy.2", {"action": "delete"}),
                ]

                responses = await opa_client.batch_evaluate_policies(evaluations)

            assert len(responses) == 3
            for response in responses:
                assert response.decision is True

    @pytest.mark.asyncio
    async def test_list_policies(self, opa_client):
        """Test listing available policies"""
        response_data = {"result": {"sdlc": {"auth": {}, "data": {}, "dlp": {}}}}

        with aioresponses() as m:
            m.get("http://localhost:8181/v1/data", payload=response_data, status=200)

            async with opa_client:
                policies = await opa_client.list_policies()

            assert "sdlc" in policies

    @pytest.mark.asyncio
    async def test_get_policy_info(self, opa_client):
        """Test getting policy information"""
        response_data = {"result": {"allow": True, "rules": ["rule1", "rule2"]}}

        with aioresponses() as m:
            m.get(
                "http://localhost:8181/v1/data/test.policy",
                payload=response_data,
                status=200,
            )

            async with opa_client:
                info = await opa_client.get_policy_info("test.policy")

            assert "result" in info
            assert info["result"]["allow"] is True

    @pytest.mark.asyncio
    async def test_health_check_success(self, opa_client):
        """Test successful health check"""
        with aioresponses() as m:
            m.get("http://localhost:8181/health", payload={"status": "ok"}, status=200)

            async with opa_client:
                is_healthy = await opa_client.health_check()

            assert is_healthy is True

    @pytest.mark.asyncio
    async def test_health_check_failure(self, opa_client):
        """Test failed health check"""
        with aioresponses() as m:
            m.get("http://localhost:8181/health", status=503)

            async with opa_client:
                is_healthy = await opa_client.health_check()

            assert is_healthy is False

    @pytest.mark.asyncio
    async def test_caching_enabled(self, opa_client, mock_redis, sample_opa_response):
        """Test decision caching"""
        # Mock cache miss, then cache set
        mock_redis.get.return_value = None

        with aioresponses() as m:
            m.post(
                "http://localhost:8181/v1/data/test.policy",
                payload=sample_opa_response,
                status=200,
            )

            async with opa_client:
                response = await opa_client.evaluate_policy(
                    "test.policy", {"user": "test"}
                )

            assert response.decision is True
            # Verify cache was set
            mock_redis.setex.assert_called_once()

    @pytest.mark.asyncio
    async def test_caching_hit(self, opa_client, mock_redis):
        """Test cache hit"""
        # Mock cache hit
        cached_data = {
            "decision": True,
            "reason": "Access granted",
            "result": {"allow": True},
            "cached_at": datetime.utcnow().isoformat(),
            "ttl": 30,
        }
        mock_redis.get.return_value = json.dumps(cached_data)

        async with opa_client:
            response = await opa_client.evaluate_policy("test.policy", {"user": "test"})

        assert response.decision is True
        assert response.reason == "Access granted"
        # Verify cache was checked but not set
        mock_redis.get.assert_called_once()
        mock_redis.setex.assert_not_called()

    @pytest.mark.asyncio
    async def test_caching_disabled(self, mock_redis, sample_opa_response):
        """Test evaluation with caching disabled"""
        config = OPAConfig(cache_enabled=False)
        client = OPAClient(config=config, redis_client=mock_redis)

        with aioresponses() as m:
            m.post(
                "http://localhost:8181/v1/data/test.policy",
                payload=sample_opa_response,
                status=200,
            )

            async with client:
                response = await client.evaluate_policy("test.policy", {"user": "test"})

            assert response.decision is True
            # Verify cache was not used
            mock_redis.get.assert_not_called()
            mock_redis.setex.assert_not_called()

    @pytest.mark.asyncio
    async def test_context_manager(self, opa_client):
        """Test async context manager"""
        async with opa_client as client:
            assert client is opa_client
            assert client.session is not None
            assert not client.session.closed

        # Session should be closed after context exit
        assert client.session.closed

    @pytest.mark.asyncio
    async def test_create_opa_client(self, mock_redis):
        """Test OPA client factory function"""
        config = OPAConfig(base_url="http://test:8181")

        with patch("redis.asyncio.from_url", return_value=mock_redis):
            client = await create_opa_client(
                config=config, redis_url="redis://localhost:6379"
            )

        assert client.config.base_url == "http://test:8181"
        assert client.redis is mock_redis


# Integration tests (would require real OPA instance)
@pytest.mark.integration
class TestOPAClientIntegration:
    """Integration tests with real OPA instance"""

    @pytest.fixture
    def opa_config(self):
        """Configuration for integration tests"""
        return OPAConfig(
            base_url="http://localhost:8181",
            timeout=5,
            cache_enabled=False,  # Disable caching for integration tests
        )

    @pytest.mark.asyncio
    async def test_real_opa_evaluation(self, opa_config):
        """Test evaluation against real OPA instance"""
        pytest.skip("Requires real OPA instance")

        client = OPAClient(config=opa_config)

        async with client:
            # Test health first
            if not await client.health_check():
                pytest.skip("OPA not available")

            # Test simple policy evaluation
            response = await client.evaluate_policy(
                "example.allow", {"user": "alice", "action": "read", "resource": "data"}
            )

            assert isinstance(response, PolicyEvaluationResponse)


# Performance tests
@pytest.mark.performance
class TestOPAClientPerformance:
    """Performance tests for OPA client"""

    @pytest.mark.asyncio
    async def test_concurrent_evaluations(self, opa_client):
        """Test concurrent policy evaluations"""
        pytest.skip("Performance test - run manually")

        async with opa_client:
            # Mock many concurrent requests
            with aioresponses() as m:
                m.post(
                    "http://localhost:8181/v1/data/test.policy",
                    payload={"result": {"allow": True}},
                    status=200,
                )

                start_time = asyncio.get_event_loop().time()

                tasks = [
                    opa_client.evaluate_policy("test.policy", {"user": f"user{i}"})
                    for i in range(100)
                ]

                responses = await asyncio.gather(*tasks)

                end_time = asyncio.get_event_loop().time()
                duration = end_time - start_time

                assert len(responses) == 100
                assert all(r.decision for r in responses)
                assert duration < 5.0  # Should complete within 5 seconds

                print(f"100 concurrent evaluations completed in {duration:.2f}s")

File: products/sdlc-ai/services/gateway/internal/infrastructure/opa/opa_client_test.go
package opa

import (
	"bytes"
	"context"
	"encoding/json"
	"io"
	"net/http"
	"strings"
	"testing"
	"time"

	"github.com/go-redis/redis/v8"
	"github.com/google/uuid"
	"github.com/sirupsen/logrus"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/mock"
	"github.com/stretchr/testify/require"

	"github.com/sdlc-ai/platform/services/gateway/internal/domain/models"
)

// MockRedisClient is a mock for Redis client
type MockRedisClient struct {
	mock.Mock
}

func (m *MockRedisClient) Get(ctx context.Context, key string) *redis.StringCmd {
	args := m.Called(ctx, key)
	cmd := redis.NewStringCmd(ctx)
	if args.Get(0) != nil {
		cmd.SetVal(args.String(0))
	}
	if args.Get(1) != nil {
		cmd.SetErr(args.Error(1))
	}
	return cmd
}

func (m *MockRedisClient) Set(ctx context.Context, key string, value interface{}, expiration time.Duration) *redis.StatusCmd {
	args := m.Called(ctx, key, value, expiration)
	cmd := redis.NewStatusCmd(ctx)
	if args.Get(0) != nil {
		cmd.SetVal(args.String(0))
	}
	if args.Get(1) != nil {
		cmd.SetErr(args.Error(1))
	}
	return cmd
}

func (m *MockRedisClient) Del(ctx context.Context, keys ...string) *redis.IntCmd {
	args := m.Called(ctx, keys)
	cmd := redis.NewIntCmd(ctx)
	if args.Get(0) != nil {
		cmd.SetVal(args.Int(0))
	}
	if args.Get(1) != nil {
		cmd.SetErr(args.Error(1))
	}
	return cmd
}

// MockRoundTripper is a mock HTTP transport
type MockRoundTripper struct {
	mock.Mock
}

func (m *MockRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
	args := m.Called(req)
	if args.Get(0) == nil {
		return nil, args.Error(1)
	}
	return args.Get(0).(*http.Response), args.Error(1)
}

func TestNewOPAClient(t *testing.T) {
	config := DefaultOPAConfig()
	logger := logrus.New()

	client, err := NewOPAClient(config, nil, logger)
	require.NoError(t, err)
	require.NotNil(t, client)

	assert.Equal(t, config.BaseURL, client.baseURL)
	assert.NotNil(t, client.httpClient)
	assert.NotNil(t, client.logger)
}

func TestOPAClient_EvaluatePolicy(t *testing.T) {
	// Create mock HTTP transport
	mockTransport := &MockRoundTripper{}

	// Mock OPA response
	opaResponse := map[string]interface{}{
		"result": map[string]interface{}{
			"allow":           true,
			"decision_reason": []string{"Access granted"},
		},
	}

	responseBody, _ := json.Marshal(opaResponse)

	mockTransport.On("RoundTrip", mock.AnythingOfType("*http.Request")).Return(&http.Response{
		StatusCode: http.StatusOK,
		Body:       io.NopCloser(bytes.NewReader(responseBody)),
		Header:     make(http.Header),
	}, nil)

	// Create client with mock transport
	config := DefaultOPAConfig()
	logger := logrus.New()
	logger.SetLevel(logrus.ErrorLevel) // Reduce noise in tests

	client, err := NewOPAClient(config, nil, logger)
	require.NoError(t, err)

	client.httpClient.Transport = mockTransport

	// Test policy evaluation
	ctx := context.Background()
	input := map[string]interface{}{
		"user":     "test-user",
		"action":   "read",
		"resource": "test-resource",
	}

	resp, err := client.EvaluatePolicy(ctx, "sdlc.test.policy", input)
	require.NoError(t, err)
	require.NotNil(t, resp)

	assert.True(t, resp.Decision)
	assert.Equal(t, "Access granted", resp.Reason)
	mockTransport.AssertExpectations(t)
}

func TestOPAClient_EvaluateDataPolicy(t *testing.T) {
	// Create mock HTTP transport
	mockTransport := &MockRoundTripper{}

	opaResponse := map[string]interface{}{
		"result": map[string]interface{}{
			"allow": true,
		},
	}

	responseBody, _ := json.Marshal(opaResponse)

	mockTransport.On("RoundTrip", mock.AnythingOfType("*http.Request")).Return(&http.Response{
		StatusCode: http.StatusOK,
		Body:       io.NopCloser(bytes.NewReader(responseBody)),
		Header:     make(http.Header),
	}, nil)

	config := DefaultOPAConfig()
	logger := logrus.New()
	logger.SetLevel(logrus.ErrorLevel)

	client, err := NewOPAClient(config, nil, logger)
	require.NoError(t, err)

	client.httpClient.Transport = mockTransport

	// Test data policy evaluation
	ctx := context.Background()
	tenantID := uuid.New()
	userID := uuid.New()

	resp, err := client.EvaluateDataPolicy(ctx, tenantID, userID, "read", "documents", map[string]interface{}{
		"id": "doc-123",
	})
	require.NoError(t, err)
	require.NotNil(t, resp)

	assert.True(t, resp.Decision)
	mockTransport.AssertExpectations(t)
}

func TestOPAClient_EvaluateDLPPolicy(t *testing.T) {
	// Create mock HTTP transport
	mockTransport := &MockRoundTripper{}

	opaResponse := map[string]interface{}{
		"result": map[string]interface{}{
			"allow":           false,
			"decision_reason": []string{"Content blocked: contains high-risk PII"},
		},
	}

	responseBody, _ := json.Marshal(opaResponse)

	mockTransport.On("RoundTrip", mock.AnythingOfType("*http.Request")).Return(&http.Response{
		StatusCode: http.StatusOK,
		Body:       io.NopCloser(bytes.NewReader(responseBody)),
		Header:     make(http.Header),
	}, nil)

	config := DefaultOPAConfig()
	logger := logrus.New()
	logger.SetLevel(logrus.ErrorLevel)

	client, err := NewOPAClient(config, nil, logger)
	require.NoError(t, err)

	client.httpClient.Transport = mockTransport

	// Test DLP policy evaluation
	ctx := context.Background()
	content := "User's SSN is 123-45-6789"
	userContext := map[string]interface{}{
		"user_id": uuid.New().String(),
		"role":    "user",
	}

	resp, err := client.EvaluateDLPPolicy(ctx, content, userContext)
	require.NoError(t, err)
	require.NotNil(t, resp)

	assert.False(t, resp.Decision)
	assert.Contains(t, resp.Reason, "high-risk PII")
	mockTransport.AssertExpectations(t)
}

func TestOPAClient_BatchEvaluatePolicies(t *testing.T) {
	// Create mock HTTP transport
	mockTransport := &MockRoundTripper{}

	opaResponse := map[string]interface{}{
		"result": map[string]interface{}{
			"allow": true,
		},
	}

	responseBody, _ := json.Marshal(opaResponse)

	mockTransport.On("RoundTrip", mock.AnythingOfType("*http.Request")).Return(&http.Response{
		StatusCode: http.StatusOK,
		Body:       io.NopCloser(bytes.NewReader(responseBody)),
		Header:     make(http.Header),
	}, nil).Times(3)

	config := DefaultOPAConfig()
	logger := logrus.New()
	logger.SetLevel(logrus.ErrorLevel)

	client, err := NewOPAClient(config, nil, logger)
	require.NoError(t, err)

	client.httpClient.Transport = mockTransport

	// Test batch evaluation
	ctx := context.Background()
	evaluations := []BatchEvaluation{
		{
			PolicyPath: "sdlc.test.policy1",
			Input:      map[string]interface{}{"action": "read"},
		},
		{
			PolicyPath: "sdlc.test.policy2",
			Input:      map[string]interface{}{"action": "write"},
		},
		{
			PolicyPath: "sdlc.test.policy3",
			Input:      map[string]interface{}{"action": "delete"},
		},
	}

	responses, err := client.BatchEvaluatePolicies(ctx, evaluations)
	require.NoError(t, err)
	require.Len(t, responses, 3)

	for _, resp := range responses {
		assert.True(t, resp.Decision)
	}

	mockTransport.AssertExpectations(t)
}

func TestOPAClient_Caching(t *testing.T) {
	// Create mock Redis client
	mockRedis := &MockRedisClient{}

	// Mock cache miss
	mockRedis.On("Get", mock.Anything, mock.AnythingOfType("string")).Return("", redis.Nil)
	mockRedis.On("Set", mock.Anything, mock.AnythingOfType("string"), mock.AnythingOfType("time.Duration")).Return("OK", nil)

	// Create mock HTTP transport
	mockTransport := &MockRoundTripper{}

	opaResponse := map[string]interface{}{
		"result": map[string]interface{}{
			"allow": true,
		},
	}

	responseBody, _ := json.Marshal(opaResponse)

	mockTransport.On("RoundTrip", mock.AnythingOfType("*http.Request")).Return(&http.Response{
		StatusCode: http.StatusOK,
		Body:       io.NopCloser(bytes.NewReader(responseBody)),
		Header:     make(http.Header),
	}, nil)

	config := DefaultOPAConfig()
	config.CacheEnabled = true
	logger := logrus.New()
	logger.SetLevel(logrus.ErrorLevel)

	client, err := NewOPAClient(config, nil, logger)
	require.NoError(t, err)

	client.httpClient.Transport = mockTransport
	client.redis = mockRedis

	// Test caching
	ctx := context.Background()
	input := map[string]interface{}{
		"user":   "test-user",
		"action": "read",
	}

	resp, err := client.EvaluatePolicy(ctx, "sdlc.test.policy", input)
	require.NoError(t, err)
	require.NotNil(t, resp)

	assert.True(t, resp.Decision)
	mockRedis.AssertExpectations(t)
	mockTransport.AssertExpectations(t)
}

func TestOPAClient_HealthCheck(t *testing.T) {
	// Create mock HTTP transport
	mockTransport := &MockRoundTripper{}

	mockTransport.On("RoundTrip", mock.AnythingOfType("*http.Request")).Return(&http.Response{
		StatusCode: http.StatusOK,
		Body:       io.NopCloser(strings.NewReader("OK")),
		Header:     make(http.Header),
	}, nil)

	config := DefaultOPAConfig()
	logger := logrus.New()
	logger.SetLevel(logrus.ErrorLevel)

	client, err := NewOPAClient(config, nil, logger)
	require.NoError(t, err)

	client.httpClient.Transport = mockTransport

	// Test health check
	ctx := context.Background()
	err = client.HealthCheck(ctx)
	require.NoError(t, err)

	mockTransport.AssertExpectations(t)
}

func TestOPAClient_HealthCheck_Failure(t *testing.T) {
	// Create mock HTTP transport
	mockTransport := &MockRoundTripper{}

	mockTransport.On("RoundTrip", mock.AnythingOfType("*http.Request")).Return(&http.Response{
		StatusCode: http.StatusServiceUnavailable,
		Body:       io.NopCloser(strings.NewReader("Service Unavailable")),
		Header:     make(http.Header),
	}, nil)

	config := DefaultOPAConfig()
	logger := logrus.New()
	logger.SetLevel(logrus.ErrorLevel)

	client, err := NewOPAClient(config, nil, logger)
	require.NoError(t, err)

	client.httpClient.Transport = mockTransport

	// Test health check failure
	ctx := context.Background()
	err = client.HealthCheck(ctx)
	require.Error(t, err)
	assert.Contains(t, err.Error(), "status 503")

	mockTransport.AssertExpectations(t)
}

func TestOPAClient_EvaluateResourcePolicy(t *testing.T) {
	// Create mock HTTP transport
	mockTransport := &MockRoundTripper{}

	opaResponse := map[string]interface{}{
		"result": map[string]interface{}{
			"allow":           true,
			"decision_reason": []string{"Resource access granted"},
		},
	}

	responseBody, _ := json.Marshal(opaResponse)

	mockTransport.On("RoundTrip", mock.AnythingOfType("*http.Request")).Return(&http.Response{
		StatusCode: http.StatusOK,
		Body:       io.NopCloser(bytes.NewReader(responseBody)),
		Header:     make(http.Header),
	}, nil)

	config := DefaultOPAConfig()
	logger := logrus.New()
	logger.SetLevel(logrus.ErrorLevel)

	client, err := NewOPAClient(config, nil, logger)
	require.NoError(t, err)

	client.httpClient.Transport = mockTransport

	// Test resource policy evaluation
	ctx := context.Background()
	policy := &models.Policy{
		ID:       uuid.New(),
		TenantID: uuid.New(),
		Name:     "Test Policy",
		Type:     models.PolicyTypeData,
		RegoPolicy: `package test.policy
default allow = false
allow {
    input.action == "read"
}`,
		Version: 1,
	}

	input := map[string]interface{}{
		"action": "read",
		"user":   "test-user",
	}

	resp, err := client.EvaluateResourcePolicy(ctx, policy, input)
	require.NoError(t, err)
	require.NotNil(t, resp)

	assert.True(t, resp.Decision)
	assert.Equal(t, "Resource access granted", resp.Reason)
	mockTransport.AssertExpectations(t)
}

func TestOPAClient_ListPolicies(t *testing.T) {
	// Create mock HTTP transport
	mockTransport := &MockRoundTripper{}

	opaResponse := map[string]interface{}{
		"result": map[string]interface{}{
			"sdlc": map[string]interface{}{
				"auth": map[string]interface{}{},
				"data": map[string]interface{}{},
				"dlp":  map[string]interface{}{},
			},
		},
	}

	responseBody, _ := json.Marshal(opaResponse)

	mockTransport.On("RoundTrip", mock.AnythingOfType("*http.Request")).Return(&http.Response{
		StatusCode: http.StatusOK,
		Body:       io.NopCloser(bytes.NewReader(responseBody)),
		Header:     make(http.Header),
	}, nil)

	config := DefaultOPAConfig()
	logger := logrus.New()
	logger.SetLevel(logrus.ErrorLevel)

	client, err := NewOPAClient(config, nil, logger)
	require.NoError(t, err)

	client.httpClient.Transport = mockTransport

	// Test list policies
	ctx := context.Background()
	policies, err := client.ListPolicies(ctx)
	require.NoError(t, err)
	require.NotNil(t, policies)

	assert.Contains(t, policies, "sdlc")
	mockTransport.AssertExpectations(t)
}

func TestOPAClient_GetPolicyInfo(t *testing.T) {
	// Create mock HTTP transport
	mockTransport := &MockRoundTripper{}

	opaResponse := map[string]interface{}{
		"result": map[string]interface{}{
			"allow": true,
			"rules": []string{"test_rule_1", "test_rule_2"},
		},
	}

	responseBody, _ := json.Marshal(opaResponse)

	mockTransport.On("RoundTrip", mock.AnythingOfType("*http.Request")).Return(&http.Response{
		StatusCode: http.StatusOK,
		Body:       io.NopCloser(bytes.NewReader(responseBody)),
		Header:     make(http.Header),
	}, nil)

	config := DefaultOPAConfig()
	logger := logrus.New()
	logger.SetLevel(logrus.ErrorLevel)

	client, err := NewOPAClient(config, nil, logger)
	require.NoError(t, err)

	client.httpClient.Transport = mockTransport

	// Test get policy info
	ctx := context.Background()
	info, err := client.GetPolicyInfo(ctx, "sdlc.test.policy")
	require.NoError(t, err)
	require.NotNil(t, info)

	assert.Contains(t, info, "result")
	mockTransport.AssertExpectations(t)
}

// Benchmark tests
func BenchmarkOPAClient_EvaluatePolicy(b *testing.B) {
	// This benchmark would require a real OPA instance
	// For now, we'll skip it
	b.Skip("Requires real OPA instance")
}

func BenchmarkOPAClient_BatchEvaluatePolicies(b *testing.B) {
	// This benchmark would require a real OPA instance
	// For now, we'll skip it
	b.Skip("Requires real OPA instance")
}
